python/lsst/pipe/base/caching_limited_butler.py

```diff
@@ -0,0 +1,194 @@
+
+    A `CachingLimitedButler` caches on both `.put()` and `.get()`, and holds a
+    single instance of the most recently used dataset type for that put/get.
+
+    The dataset types which will be cached on put/get are controlled via the
+    `cache_on_put` and `cache_on_get` attributes, respectively.
+
+    By default, copies of the cached items are returned on `get`, so that code
+    is free to operate on data in-place. A `no_copy_on_cache` attribute also
+    exists to tell the `CachingLimitedButler` not to return copies when it is
+    known that the calling code can be trusted not to change values, e.g. when
+    passing calibs to `isrTask`.
+
+    Parameters
+    ----------
+    wrapped : `LimitedButler`
+        The butler to wrap.
+    cache_on_put : `set` [`str`], optional
+        The dataset types to cache on put.
+    cache_on_get : `set` [`str`], optional
+        The dataset types to cache on get.
+    no_copy_on_cache : `set` [`str`], optional
+        The dataset types for which to not return copies when cached.
+    """
+
+    def __init__(
+        self,
+        wrapped: LimitedButler,
+        cache_on_put: Set[str] = frozenset(),
+        cache_on_get: Set[str] = frozenset(),
+        no_copy_on_cache: Set[str] = frozenset(),
+    ):
+        self._wrapped = wrapped
+        self._datastore = self._wrapped._datastore
+        self.storageClasses = self._wrapped.storageClasses
+        self._cache_on_put = cache_on_put
+        self._cache_on_get = cache_on_get
+        self._cache: dict[str, tuple[DatasetId, InMemoryDatasetHandle]] = {}
+        self._no_copy_on_cache = no_copy_on_cache
+
+    def get(
+        self,
+        ref: DatasetRef,
+        /,
+        *,
+        parameters: dict[str, Any] | None = None,
+        storageClass: StorageClass | str | None = None,
+    ) -> Any:
+        if storageClass is None:
+            storageClass = ref.datasetType.storageClass
+        elif isinstance(storageClass, str):
+            storageClass = self.storageClasses.getStorageClass(storageClass)
+
+        # check if we have this dataset type in the cache
+        if cached := self._cache.get(ref.datasetType.name):
+            dataset_id, handle = cached
+            if dataset_id == ref.id:  # if we do, check it's the right object
+                _LOG.debug("Returning cached dataset %s", ref)
+                return handle.get(parameters=parameters, storageClass=storageClass)
+
+        obj = self._wrapped.get(ref, parameters=parameters, storageClass=storageClass)
+        if ref.datasetType.name in self._cache_on_get and not parameters:
+            # and not parameters is to make sure we don't cache sub-images etc
+            self._cache[ref.datasetType.name] = (
+                ref.id,
+                InMemoryDatasetHandle(
+                    obj,
+                    storageClass=storageClass,
+                    dataId=ref.dataId,
+                    copy=ref.datasetType.name not in self._no_copy_on_cache,
```

Comment:
I understand that `copy` flag only matters on `handle.get`? So if you save this handle and return `obj`, the task could modify `obj` which will also modify contents of the `handle`? Probably need to copy `obj` before storing it in handle (unless `copy` is false)?

Reply:
I'm actually not quite sure. @TallJimbo could you comment, do you think?

Reply:
We resolved this OOB. Changes pushed and Jenkins running.